c10/core/TensorImpl.h

```diff
@@ -1401,6 +1397,17 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   // agree with the type meta in storage
   caffe2::TypeMeta data_type_;
 
+  // NOTE [c10::optional operator usage in CUDA]
+  // Our optional definition doesn't compile in .cu file if `value()` or
+  // `operator->` are used.  Instead, we always use `operator*`.
+  // See https://github.com/pytorch/pytorch/issues/18496 for more info.
+  // If this is too burdensome to maintain, we can just
+  // manually implement this with an additional bool.
+
+  // INVARIANT: When storage is non-null, this Device must
+  // agree with the type meta in storage.
+  c10::optional<c10::Device> device_opt_;
```

Comment:
why is it optional even? for any tensor except Undefined one it should initialized. For Undefined we can just populate fake value from DeviceType enum, e.g. ONLY_FOR_TEST.

My worry is that c10::optional creates visibility that it's not always populated, though for well-formed tensor it actually is

Reply:
I considered that but it seems worse.  IMO adding an ONLY_FOR_TEST is first of all misleading because it's not only for test, and it moves the complexity from where it should be (when the device really can be non-existent, here) to where it shouldn't be (like now everywhere else in code, where the device must exist).

IMO the correct solution is to:
1) add the interface as above to get the actual device without allocating something.
2) hook up SparseTensor to 1) so the device is properly set in the constructor
3) assert that only UndefinedTensors don't have a devices in the constructor.

Reply:
The basic question is whether or not `Device` should be default constructible or not. We've gotten away with it not being default constructible, and the type-safe person in me likes it.